### PR TITLE
Wingloss

### PR DESCRIPTION
**Pull request by jin-s13**

### CODE REVIEW

File: configs/face/deeppose/wflw/deeppose_res50_wflw_256x256.py [hunk 1]
```diff
@@ -46,11 +46,7 @@
         num_joints=channel_cfg['num_output_channels'],
         loss_keypoint=dict(type='SmoothL1Loss', use_target_weight=True)),
     train_cfg=dict(),
-    test_cfg=dict(
-        flip_test=True,
-        post_process='default',
-        shift_heatmap=True,
-        modulate_kernel=11))
+    test_cfg=dict())
```
Comment (innerlee) [inline configs/face/deeppose/wflw/deeppose_res50_wflw_256x256.py:49]: Reason and effect of this change?

File: mmpose/models/losses/regression_loss.py [hunk 2]
```diff
@@ -20,18 +29,72 @@ def forward(self, output, target, target_weight):
...
+                 loss_weight=1.):
+        super().__init__()
+        self.omega = omega
+        self.epsilon = epsilon
+        self.use_target_weight = use_target_weight
+        self.loss_weight = loss_weight
+
+    def criterion(self, pred, target):
+        delta = (target - pred).abs()
+        c = self.omega * (1.0 - math.log(1.0 + self.omega / self.epsilon))
```
Comment (innerlee) [inline mmpose/models/losses/regression_loss.py:74]: It seems that `c` could be pre-computed

File: mmpose/models/losses/regression_loss.py [hunk 3]
```diff
@@ -20,18 +29,72 @@ def forward(self, output, target, target_weight):
...
+        losses = torch.where(
+            torch.lt(delta, self.omega),
+            self.omega * torch.log(1.0 + delta / self.epsilon), delta - c)
```
Comment (innerlee) [inline mmpose/models/losses/regression_loss.py:77]: `self.omega * torch.log(1.0 + delta / self.epsilon)` could be cached also

File: mmpose/models/losses/regression_loss.py [hunk 4]
```diff
@@ -20,18 +29,72 @@ def forward(self, output, target, target_weight):
...
+        self.omega = omega
+        self.epsilon = epsilon
+        self.use_target_weight = use_target_weight
+        self.loss_weight = loss_weight
+
+    def criterion(self, pred, target):
+        delta = (target - pred).abs()
+        c = self.omega * (1.0 - math.log(1.0 + self.omega / self.epsilon))
+        losses = torch.where(
+            torch.lt(delta, self.omega),
```
Comment (innerlee) [inline mmpose/models/losses/regression_loss.py:76]: use `<` instead of `lt`

File: mmpose/models/losses/regression_loss.py [hunk 5]
```diff
@@ -20,18 +29,72 @@ def forward(self, output, target, target_weight):
...
+            target_weight (torch.Tensor[N, K, D]):
+                Weights across different joint types.
+        """
+        if self.use_target_weight:
+            loss = self.criterion(output * target_weight,
+                                  target * target_weight)
+        else:
+            loss = self.criterion(output, target)
+
+        return loss * self.loss_weight
```
Comment (innerlee) [inline mmpose/models/losses/regression_loss.py:46]: Looks like that the implementation of `SmoothL1Loss` is changed.

Why remove the ` / num_joints` part?

File: configs/face/deeppose/wflw/deeppose_res50_wflw_256x256.py (hunk 1 above)
Comment (jin-s13) [reply configs/face/deeppose/wflw/deeppose_res50_wflw_256x256.py:49]: They are not used in fc_head.

File: mmpose/models/losses/regression_loss.py (hunk 5 above)
Comment (jin-s13) [reply mmpose/models/losses/regression_loss.py:46]: Yes. 
In nn.SmoothL1Loss, reduction='mean', so the loss has already been divided by the number of elements in the output. No need to divide `num_joints` again.

Comment (innerlee) [reply mmpose/models/losses/regression_loss.py:46]: Okay, does this change affect any ckpt?

Comment (jin-s13) [reply mmpose/models/losses/regression_loss.py:46]: Yes. This may affect the ckpts of deeppose.

File: mmpose/models/losses/regression_loss.py (hunk 2 above)
Comment (jin-s13) [reply mmpose/models/losses/regression_loss.py:74]: Yes.

File: tests/test_loss/test_regression_losses.py [hunk 6]
```diff
@@ -63,3 +63,6 @@ def test_wing_loss():
     assert torch.gt(
         loss(fake_pred, fake_label, torch.ones_like(fake_label)),
         torch.tensor(.5))
+
+
+test_wing_loss()
```
Comment (innerlee) [inline tests/test_loss/test_regression_losses.py:68]: This line is not needed in the final version.

BTW for better experience in debugging unittest, try vscode. 

image

File: mmpose/models/losses/regression_loss.py (hunk 5 above)
Comment (jin-s13) [reply mmpose/models/losses/regression_loss.py:46]: I retrained the model, but did not notice significant difference in accuracy.

**Review by innerlee — APPROVED** [review]
